### ISSUE TITLE
Investigate WebGL support

### ISSUE THREAD

**Issue by samreid** [phetsims/forces-and-motion-basics#103]
Performance is still a bit sluggish on firefox and iPad. Can this be improved with WebGL?

**Comment by samreid** [comment]
I tried renderer:'webgl' on a few of the items and didn't see any change in performance.  Putting 'webgl' for all of the items in the motion screen causes "too many webgl contexts" error, probably because it is alternating image/text/image/text/etc.

**Comment by samreid** [comment]
I'll look into this for S2015R.  Unassigning until then.

**Comment by jessegreenberg** [comment]
Sim is very smooth on my Windows 10 machine with Firefox.  Will check on other test devices tomorrow.

**Comment by jessegreenberg** [comment]
Just tested on an iPad, can confirm that the sim is just a little slugish.  WebGL might improve performance here.

**Comment by jessegreenberg** [comment]
@samreid for this issue, do you envision a re-implementation using WebGL (in a similar way to Energy Skate Park: Basics) or is this issue more to investigate renderer settings?

**Comment by samreid** [comment]
Energy Skate Park: Basics took the following WebGL strategy:

1. Move all WebGL content to the same layer (generally the front), this meant splitting up background/foreground of the bar chart, and moving the skater to the front, etc.  You can use the scenery popupDebug to see if you've got everything into a single WebGL layer (or 2 if necessary).  This pertains to the issue described in https://github.com/phetsims/forces-and-motion-basics/issues/103#issuecomment-112886617

2. Converted some shapes to images using toCanvasNodeSynchronous().  This is done for the dot under the skater as well as the GaugeNode needle.

3. Specify renderer:'webgl' for the images and rectangles.  In ESPB these are the bar chart rectangles and the skater images.

4. Implemented a custom WebGLNode subclass for the PieChartWebGLNode.

For Forces and Motion: Basics, we should investigate using renderer:'webgl', which will probably involve steps 1-3 above.  WebGLNode subclasses hopefully will not be needed here.

> Just tested on an iPad, can confirm that the sim is just a little sluggish.

Steps 1-3 above will be time consuming but also generally detract from the maintainability of the simulation since nodes that naturally belong together (e.g. images and their mass readouts) would have to be somehow separated.  The long-term solution to this issue would be to get more and more scenery nodes (such as text nodes) the ability to render in WebGL.  You could do a performance feasibility test by putting renderer:'webgl' on the images and commenting out (temporarily) the text nodes.  My recommendation: OK to investigate the performance/time/maintainability tradeoffs and see what would be entailed but should coordinate with @ariel-phet before spending a lot of time on this (especially if it is *nearly* good enough on iPad already).

**Comment by samreid** [comment]
Another point: we are generally proceeding with quick fixes for S2015R as long as the performance is equivalent to the previously published version.  Due to the amount of work described in the preceding comment, this may make more sense to do S2015R without it to make the i18n version available asap, then leaving WebGL for future work.

**Comment by jessegreenberg** [comment]
Excellent, thank you for the insight @samreid!  These steps make sense.  I discussed this on 3/3/16 with @ariel-phet and he wanted a time scale for this issue to understand whether it should be completed before the redeploy or after.  https://github.com/phetsims/forces-and-motion-basics/issues/103#issuecomment-193363279 helps put things in perspective and lays out a plan for the WebGL work.

From the recommendations above, I am also inclined to proceed with the redeploy so we can get it out quickly.  This issue could be continued after the redeploy is complete.  @ariel-phet does this plan sound good to you?

**Comment by ariel-phet** [comment]
@jessegreenberg postponing until after redeploy sounds good. Once the redeploys are finished, lets revisit this issue and see where it falls in your todo list.

Removing S2015R label, marking as priority deferred

**Comment by jessegreenberg** [comment]
Unassigning myself from this issue, there is no intention to work on this in the near future.

**Comment by samreid** [comment]
I published a dev version from master, @KatieWoe can you please check performance on our slowest supported platform (which I think is iPad Air 2)? 

https://phet-dev.colorado.edu/html/forces-and-motion-basics/2.4.0-dev.3/phet/forces-and-motion-basics_en_phet.html

**Comment by KatieWoe** [comment]
Looks ok on iOS 11 and Win 7 IE.  iOS 11 had some FPS drops on profiler, but I didn't see anything when using it.

**Comment by samreid** [comment]
Thanks @KatieWoe, perhaps no further work is necessary for graphics performance optimization.  Closing.
